### PR TITLE
Optimize thread usage and handle http exceptions. 

### PR DESCRIPTION
**Pull request by vilyapilya**
Most of the changes are in the client file. I removed the scheduling library bc it did not do the timing. 
I also changed the names of some variable, which we will probably need to evert to the old ones if if we have a lot customer using this library.

### CODE REVIEW

File: lib/logdna/resources.rb [hunk 1]
```diff
@@ -6,8 +6,9 @@ module Resources
   MAX_REQUEST_TIMEOUT = 300000
   MAX_LINE_LENGTH = 32000
   MAX_INPUT_LENGTH = 80
+  BACKOFF_PERIOD = 60
   FLUSH_INTERVAL = 0.25
-  TIMER_OUT = 15
+  TIMER_OUT = 120
```
Comment (smusali) [inline lib/logdna/resources.rb]: Why `120` instead of `30` and `TIMER_OUT` instead of `TIMEOUT` or `REQUEST_TIMEOUT`?

Comment (vilyapilya) [reply lib/logdna/resources.rb]: > Let's try to make it consistent with [`Node.js Code Library`](https://github.com/logdna/nodejs) regarding the configurable options...

@smusali  yes I had this idea but I am not sure how safe it is to change the names of the configurable options. If we change it, it make break the users' code.

Comment (vilyapilya) [reply lib/logdna/resources.rb]: As you can see, I did not give the name to this variable.

File: lib/logdna.rb [hunk 2]
```diff
@@ -20,50 +20,24 @@ def initialize(key, opts={})
...
       @meta = opts[:meta]
-      @@client = nil unless defined? @@client
-      
+
       endpoint = opts[:endpoint] || Resources::ENDPOINT
       hostname = opts[:hostname] || Socket.gethostname
       ip =  opts.key?(:ip) ? "&ip=#{opts[:ip]}" : ''
       mac = opts.key?(:mac) ? "&mac=#{opts[:mac]}" : ''
       url = "#{endpoint}?hostname=#{hostname}#{mac}#{ip}"
+      uri = URI(url)
```
Comment (dchai76) [inline lib/logdna.rb]: Looks like you're no longer handling `URI::ValidURIRequired` - is that OK?

File: lib/logdna.rb [hunk 3]
```diff
@@ -20,50 +20,24 @@ def initialize(key, opts={})
...
-      begin
-        if (hostname.size > Resources::MAX_INPUT_LENGTH || @app.size > Resources::MAX_INPUT_LENGTH )
-            raise MaxLengthExceeded.new
-        end
-      rescue MaxLengthExceeded => e
+      if (hostname.size > Resources::MAX_INPUT_LENGTH || @app.size > Resources::MAX_INPUT_LENGTH )
         puts "Hostname or Appname is over #{Resources::MAX_INPUT_LENGTH} characters"
-        handle_exception(e)
-        return
       end
```
Comment (dchai76) [inline lib/logdna.rb]: This used to early return and doesn't anymore - is that intentional?

File: lib/logdna.rb [hunk 4]
```diff
@@ -20,50 +20,24 @@ def initialize(key, opts={})
...
-
-      begin
-        request = Net::HTTP::Post.new(uri.request_uri, 'Content-Type' => 'application/json')
-        request.basic_auth 'username', key
-      rescue => e
-        handle_exception(e)
-        return
-      end
+      request = Net::HTTP::Post.new(uri.request_uri, 'Content-Type' => 'application/json')
+      request.basic_auth 'username', key
```
Comment (dchai76) [inline lib/logdna.rb]: I'd prefer we be consistent with syntax and use parens everywhere: `request.basic_auth('username', key)`

File: lib/logdna.rb [hunk 5]
```diff
@@ -20,50 +20,24 @@ def initialize(key, opts={})
...
-      end
-
-      begin
-        request = Net::HTTP::Post.new(uri.request_uri, 'Content-Type' => 'application/json')
-        request.basic_auth 'username', key
-      rescue => e
-        handle_exception(e)
-        return
-      end
+      request = Net::HTTP::Post.new(uri.request_uri, 'Content-Type' => 'application/json')
```
Comment (dchai76) [inline lib/logdna.rb]: We're no longer handling or logging exceptions for this call - intentional?

File: lib/logdna.rb [hunk 6]
```diff
@@ -82,14 +56,16 @@ def level=(value)
       @level = value
     end
 
-    def log(msg=nil, opts={})
-      loggerExist?
-      message = msg
-      message = yield if msg.nil? && block_given?
-      @response = @@client.buffer(message, default_opts.merge(opts).merge({
+    def log(message, opts={})
+      if(message.length == 0)
```
Comment (dchai76) [inline lib/logdna.rb]: (nit) space after `if`. And I'm more accustomed to using `if (message.empty?)`

File: lib/logdna.rb [hunk 7]
```diff
@@ -82,14 +56,16 @@ def level=(value)
...
+        puts "Your logline cannot be empty"
+        return
+      end
+      message = message.to_s unless message.instance_of? String
```
Comment (dchai76) [inline lib/logdna.rb]: I'd drop the whole `unless` clause.

File: lib/logdna.rb [hunk 8]
```diff
@@ -82,14 +56,16 @@ def level=(value)
...
+      message = message.encode("UTF-8")
+      @@client.write_to_buffer(message, default_opts.merge(opts).merge({
```
Comment (dchai76) [inline lib/logdna.rb]: So it looks like you no longer check if `@@client` exists (when authorization fails). Is that a problem?

File: lib/logdna.rb [hunk 9]
```diff
@@ -142,18 +111,5 @@ def datetime_format(*arg)
       puts "datetime_format not supported in LogDNA logger"
       return false
     end
-
-
-    def close
```
Comment (dchai76) [inline lib/logdna.rb]: Can you explain why you removed these?

File: lib/logdna/client.rb [hunk 10]
```diff
@@ -11,153 +12,114 @@ def initialize(request, uri, opts)
...
 
-    def exitout()
-      check_side_buffer
-      if @messages.any?
+   def exitout
+      if @buffer.any?
         flush()
       end
-      puts "Exiting LogDNA logger: Logging remaining messages"
+        puts "Exiting LogDNA logger: Logging remaining messages"
```
Comment (dchai76) [inline lib/logdna/client.rb]: This indentation looks wrong. Did you lint this file? Try Rubocop if you haven't.

File: lib/logdna/client.rb [hunk 11]
```diff
@@ -11,153 +12,114 @@ def initialize(request, uri, opts)
...
 
       @lock = Mutex.new
-      @task = nil
-
-      # NOTE: the byte limit only affects the message, not the entire message_hash
-      @actual_byte_limit = opts[:flushbyte] ||= Resources::FLUSH_BYTE_LIMIT
-      @actual_flush_interval = opts[:flushtime] ||= Resources::FLUSH_INTERVAL
-
-      @@request = request
+      @flush_limit = opts[:flush_size] ||= Resources::FLUSH_BYTE_LIMIT
```
Comment (dchai76) [inline lib/logdna/client.rb]: I'm pretty sure we're misusing `||=` here. That operator sets the variable if it's falsy. So what we're doing here is setting `opts[:flush_size]` to `Resources::FLUSH_BYTE_LIMIT` if it's not already set, then setting `@flush_limit` to that value. I don't see the `opts` being used again, I think we just want to set `@flush_limit`, so `@flush_limit = opts[:flush_size] ||= Resources::FLUSH_BYTE_LIMIT`. Same below in line 22 and anywhere else we're using `||=`.

File: lib/logdna/client.rb [hunk 12]
```diff
@@ -11,153 +12,114 @@ def initialize(request, uri, opts)
...
-        if @messages.any?
-          # keep running if there are queued messages, but don't flush
-          # because the buffer is being flushed due to being over the limit
-          unless @buffer_over_limit
-            flush()
-          end
-        else
-          # no messages means we can kill the task
-          task.kill
+        timer_task = Concurrent::TimerTask.new(execution_interval: @flush_interval, timeout_interval: Resources::TIMER_OUT) do |task|
```
Comment (dchai76) [inline lib/logdna/client.rb]: Indent is off here - it's 2 spaces everywhere else, 4 spaces here.

File: lib/logdna/client.rb [hunk 13]
```diff
@@ -11,153 +12,114 @@ def initialize(request, uri, opts)
...
+            puts 'executing'
```
Comment (dchai76) [inline lib/logdna/client.rb]: This looks like a stray debug message?

File: lib/logdna/client.rb [hunk 14]
```diff
@@ -11,153 +12,114 @@ def initialize(request, uri, opts)
...
-      @messages.push(message_hash(msg, opts))
-      buffer_size
-    end
-
-    def queue_to_buffer(queue=@queue)
-      next_object = queue.shift
-      write_to_buffer(next_object[:msg], next_object[:opts])
-    end
+      if @lock.try_lock
+          if !@side_messages.empty?
```
Comment (dchai76) [inline lib/logdna/client.rb]: Same note about indent, it's inconsistent where you're using 2 or 4

File: lib/logdna/resources.rb (hunk 1 above)
Comment (smusali) [reply lib/logdna/resources.rb]: `TIMER_OUT` is not such variable that customer touches while using but in general, you are right

File: lib/logdna/client.rb [hunk 15]
```diff
@@ -11,153 +12,114 @@ def initialize(request, uri, opts)
...
 
-        # don't kill @task if this was executed from self.buffer
-        # don't kill @task if there are queued messages
-        unless @buffer_over_limit || @messages.any? || @task.nil?
-          @task.shutdown
-          @task.kill
+        begin
+          @lock.unlock
+        rescue
+          puts 'Nothing was locked'
```
Comment (dchai76) [inline lib/logdna/client.rb]: This seems not too useful

File: lib/logdna.rb (hunk 3 above)
Comment (vilyapilya) [reply lib/logdna.rb]: The whole logic here is weird. Why to catch an exception that they throw by themselves ? =)

Comment (vilyapilya) [reply lib/logdna.rb]: And I am not even sure how useful this check is.  Do we care so much about how long the hostname they use? Why do we care about this only in Ruby code? Why don't we care then about the length of the app name, level name etc...

File: lib/logdna/client.rb [hunk 16]
```diff
@@ -11,153 +12,112 @@ def initialize(request, uri, opts)
...
-      end
-      msg
+      @request = request
+      @retry_timeout = opts[:retry_timeout] ||= Resources::RETRY_TIMEOUT
     end
 
-    def message_hash(msg, opts={})
-      obj = {
+    def process_message(msg, opts={})
+      processedMessage = {
```
Comment (dchai76) [inline lib/logdna/client.rb]: Use underscores, please.

File: lib/logdna.rb (hunk 3 above)
Comment (vilyapilya) [reply lib/logdna.rb]: Although I don't have any limitation to the message length. I can add it.

File: lib/logdna/client.rb [hunk 17]
```diff
@@ -11,153 +12,112 @@ def initialize(request, uri, opts)
...
+            http.request(@request)
+          end
 
-        # don't kill @task if this was executed from self.buffer
-        # don't kill @task if there are queued messages
-        unless @buffer_over_limit || @messages.any? || @task.nil?
-          @task.shutdown
-          @task.kill
+          if(@response.is_a?(Net::HTTPForbidden))
+            p "Please provide a valid ingestion key"
```
Comment (dchai76) [inline lib/logdna/client.rb]: Please be consistent and use `puts` everywhere

File: lib/logdna/client.rb [hunk 18]
```diff
@@ -11,153 +12,112 @@ def initialize(request, uri, opts)
...
+          @response = Net::HTTP.start(@uri.hostname, @uri.port, use_ssl: @uri.scheme == 'https') do |http|
+            http.request(@request)
+          end
 
-        # don't kill @task if this was executed from self.buffer
-        # don't kill @task if there are queued messages
-        unless @buffer_over_limit || @messages.any? || @task.nil?
-          @task.shutdown
-          @task.kill
+          if(@response.is_a?(Net::HTTPForbidden))
```
Comment (dchai76) [inline lib/logdna/client.rb]: (nit) Space after the `if`. This goes for everywhere in this block below.

File: lib/logdna.rb (hunk 2 above)
Comment (vilyapilya) [reply lib/logdna.rb]: I tested this functionality, it does not work. URI::ValidURIRequired 
And I don't see why to check this validation, if the url is not valid for any reason the exception will be thrown.

File: lib/logdna.rb (hunk 3 above)
Comment (dchai76) [reply lib/logdna.rb]: > The whole logic here is weird. Why to catch an exception that they throw by themselves ? =)

To log / print the exception. It's a little janky, but we should figure out why we weren't setting up a client in this hostname/app size case before. I get you have questions - we should get answers.

File: lib/logdna.rb [hunk 19]
```diff
@@ -20,50 +20,24 @@ def initialize(key, opts={})
...
-
-      begin
-        request = Net::HTTP::Post.new(uri.request_uri, 'Content-Type' => 'application/json')
-        request.basic_auth 'username', key
-      rescue => e
-        handle_exception(e)
-        return
-      end
+      request = Net::HTTP::Post.new(uri.request_uri, 'Content-Type' => 'application/json')
+      request.basic_auth('username', key)
```
Comment (dchai76) [inline lib/logdna.rb]: Does this call possibly raise an exception? What should we do?

File: lib/logdna.rb [hunk 20]
```diff
@@ -20,50 +20,24 @@ def initialize(key, opts={})
       @level = opts[:level] || 'INFO'
       @env = opts[:env]
       @meta = opts[:meta]
-      @@client = nil unless defined? @@client
```
Comment (dchai76) [inline lib/logdna.rb]: If you're going to remove this we might as well make this `@client` (instance var) instead of `@@client` (class var).

File: lib/logdna.rb [hunk 21]
```diff
@@ -82,14 +56,15 @@ def level=(value)
       @level = value
     end
 
-    def log(msg=nil, opts={})
-      loggerExist?
-      message = msg
-      message = yield if msg.nil? && block_given?
-      @response = @@client.buffer(message, default_opts.merge(opts).merge({
+    def log(message, opts={})
+      if (message.length == 0)
```
Comment (dchai76) [inline lib/logdna.rb]: Something like `if (message.empty?)` is more Ruby-ish.

File: lib/logdna.rb (hunk 8 above)
Comment (dchai76) [reply lib/logdna.rb]: > There is no authorization.

I'm not following - previously if `request.basic_auth` failed, `@@client` could be nil. We're not checking for that failure anymore so I'm not sure what happens now.

File: lib/logdna.rb (hunk 9 above)
Comment (dchai76) [reply lib/logdna.rb]: Could you explain that to me? What do you mean it doesn't do anything? `close` is the override for the parent class (`Logger::close`). `at_exit` is called on program exit. Are you saying `@@client.exitout()` is a no-op?

File: lib/logdna/client.rb [hunk 22]
```diff
@@ -1,164 +1,130 @@
...
-      # NOTE: the byte limit only affects the message, not the entire message_hash
-      @actual_byte_limit = opts[:flushbyte] ||= Resources::FLUSH_BYTE_LIMIT
-      @actual_flush_interval = opts[:flushtime] ||= Resources::FLUSH_INTERVAL
-
-      @@request = request
-    end
-
-    def encode_message(msg)
-      msg = msg.to_s unless msg.instance_of? String
+      @flush_limit = opts[:flush_size] ? opts[:flush_size] : Resources::FLUSH_BYTE_LIMIT
```
Comment (dchai76) [inline lib/logdna/client.rb]: Why did you change this format? `@flush_limit = opts[:flush_size] || Resources::FLUSH_BYTE_LIMIT` is more typical.

File: lib/logdna/client.rb [hunk 23]
```diff
@@ -1,164 +1,130 @@
...
-    def message_hash(msg, opts={})
-      obj = {
+    def process_message(msg, opts = {})
+      processed_message = {
         line: msg,
         app: opts[:app],
         level: opts[:level],
         env: opts[:env],
         meta: opts[:meta],
-        timestamp: Time.now.to_i,
```
Comment (dchai76) [inline lib/logdna/client.rb]: Restore this, it's intentional; coders do this so that when you add/remove additional opts it maintains the `git blame` to the commit that actually added / changed the line instead of a commit that added / removed something before / after it.

File: lib/logdna/client.rb [hunk 24]
```diff
@@ -1,164 +1,130 @@
...
       end
     end
 
-    def exitout()
-      check_side_buffer
-      if @messages.any?
-        flush()
+    def flush
+      @flush_scheduled = false
+      return if @buffer.empty?
```
Comment (dchai76) [inline lib/logdna/client.rb]: You got rid of the side buffer check - what happens if there are messages in `@side_messages`?

File: lib/logdna.rb (hunk 3 above)
Comment (vilyapilya) [reply lib/logdna.rb]: Actually we have the checks for all the parameters sent in options in nodejs lib. I'll add them here as well.

File: lib/logdna.rb (hunk 19 above)
Comment (vilyapilya) [reply lib/logdna.rb]: no, it does not.

File: lib/logdna.rb (hunk 8 above)
Comment (vilyapilya) [reply lib/logdna.rb]: This is the code library that writes a client that connects to our servers. Our servers use the ingestion keys for authenticating. (if a user has it, then he can send data to the server). Basic authentication does not make any sense here.

File: lib/logdna/client.rb (hunk 22 above)
Comment (vilyapilya) [reply lib/logdna/client.rb]: I thought this what you requested the last time.

Comment (dchai76) [reply lib/logdna/client.rb]: I was requesting we use `||` instead of `||=`.

File: lib/logdna.rb [hunk 25]
```diff
@@ -143,16 +114,9 @@ def datetime_format(*arg)
...
     end
 
-
     def close
-      if defined? @@client and !@@client.nil?
-          @@client.exitout()
-      end
-    end
-
-    at_exit do
```
Comment (dchai76) [inline lib/logdna.rb]: `at_exit` is a Ruby thing, wouldn't hurt to restore this also.

File: lib/logdna/client.rb [hunk 26]
```diff
@@ -1,164 +1,130 @@
...
-      @actual_byte_limit = opts[:flushbyte] ||= Resources::FLUSH_BYTE_LIMIT
-      @actual_flush_interval = opts[:flushtime] ||= Resources::FLUSH_INTERVAL
-
-      @@request = request
-    end
-
-    def encode_message(msg)
-      msg = msg.to_s unless msg.instance_of? String
+      @flush_limit = opts[:flush_size] ? opts[:flush_size] : Resources::FLUSH_BYTE_LIMIT
+      @flush_interval = opts[:flush_interval] ? opts[:flush_interval] : Resources::FLUSH_INTERVAL
```
Comment (dchai76) [inline lib/logdna/client.rb]: Same thing here, I think you mistook my request. Use `||` instead of `||=`. That's the semantic intention.

File: lib/logdna/client.rb [hunk 27]
```diff
@@ -1,164 +1,130 @@
...
 
-      begin
-          msg = msg.encode("UTF-8")
-      rescue Encoding::UndefinedConversionError => e
-        # NOTE: should this be raised or handled silently?
-        # raise e
-      end
-      msg
+      @request = request
+      @retry_timeout = opts[:retry_timeout] ? opts[:retry_timeout] : Resources::RETRY_TIMEOUT
```
Comment (dchai76) [inline lib/logdna/client.rb]: Same here.

File: lib/logdna/client.rb [hunk 28]
```diff
@@ -1,164 +1,130 @@
...
+      else
+        schedule_flush
       end
+    end
+
+    def exitout
+      flush if @buffer.any? || @side_messages.any?
       puts "Exiting LogDNA logger: Logging remaining messages"
-      return
+      nil
```
Comment (dchai76) [inline lib/logdna/client.rb]: Just curious why you changed this, I feel like I see empty `return` more than an ending `nil`.

File: lib/logdna/client.rb [hunk 29]
```diff
@@ -1,164 +1,130 @@
...
+
+      if @lock.try_lock
+        send_request
+      else
+        schedule_flush
       end
+    end
+
+    def exitout
+      flush if @buffer.any? || @side_messages.any?
```
Comment (dchai76) [inline lib/logdna/client.rb]: This test is already a part of `flush` so redundant; let's remove it and just let `flush` do it.

File: lib/logdna/client.rb [hunk 30]
```diff
@@ -1,164 +1,130 @@
...
+        flush
+      }
+      thread = Thread.new { start_timer }
+      thread.join
     end
 
     def write_to_buffer(msg, opts)
-      return if msg.nil?
-      msg = encode_message(msg)
+      if @lock.try_lock
```
Comment (dchai76) [inline lib/logdna/client.rb]: So the way the lock logic has changed makes me a bit nervous. Before we just wrapped in a `synchronize` block so we know we released in the right place. Now there's odd coordination going on, and I'm not sure it's correct, like you're releasing the lock before we flush (line 62) - is that right? If you're concerned about blocking what about just checking for `if @lock.locked?` and adding to `@side_messages` in that case, otherwise doing this logic in a `synchronize` block?

File: lib/logdna/client.rb [hunk 31]
```diff
@@ -1,164 +1,130 @@
...
 
-        # don't kill @task if this was executed from self.buffer
-        # don't kill @task if there are queued messages
-        unless @buffer_over_limit || @messages.any? || @task.nil?
-          @task.shutdown
-          @task.kill
+        if @response.is_a?(Net::HTTPForbidden)
+          puts "Please provide a valid ingestion key"
+        elsif !@response.is_a?(Net::HTTPSuccess)
+          puts "The response is not successful "
```
Comment (dchai76) [inline lib/logdna/client.rb]: No retry in this case?

File: lib/logdna/client.rb [hunk 32]
```diff
@@ -1,164 +1,130 @@
...
 
-    def exitout()
-      check_side_buffer
-      if @messages.any?
-        flush()
+    def flush
+      @flush_scheduled = false
+      return if @buffer.empty? && @side_messages.empty?
+
+      if @lock.try_lock
```
Comment (dchai76) [inline lib/logdna/client.rb]: Here's another case where the lock behavior is weird, locking here, but counting on `send_request` to release it - that's error-prone.

File: lib/logdna/client.rb [hunk 33]
```diff
@@ -1,164 +1,130 @@
...
-      end
+        @buffer_byte_size += new_message_size
+        @buffer.push(processed_message)
 
-      check_side_buffer
-      buffer_size = @buffer.write(msg)
-      @messages.push(message_hash(msg, opts))
-      buffer_size
-    end
+        @lock.unlock if @lock.locked?
```
Comment (dchai76) [inline lib/logdna/client.rb]: OK, having read through this all, I now see why you did the unlock here and to me that's an indication that the lock logic is not right - it really should be locked/unlocked in roughly the same block - there's a somewhat crazy coordination game going on right now that makes the logic unclear, and I think not quite right.

File: lib/logdna/client.rb (hunk 31 above)
Comment (vilyapilya) [reply lib/logdna/client.rb]: no, there is no point in retrying if the ingestion is not valid or not provided

File: lib/logdna/client.rb (hunk 32 above)
Comment (vilyapilya) [reply lib/logdna/client.rb]: ok, I agree. I just have to refactor the methods after adding the lints and requirements to the methods length. I'll take a look at my locking logic.

File: lib/logdna/client.rb (hunk 28 above)
Comment (vilyapilya) [reply lib/logdna/client.rb]: because of linting ) but I can change its settings.

File: lib/logdna.rb [hunk 34]
```diff
@@ -73,7 +50,7 @@ def default_opts
       }
     end
 
-    def level=(value)
+    def assign_level=(value)
```
Comment (dchai76) [inline lib/logdna.rb]: Undo this - I think you're misunderstanding the syntax here. This is a virtual attribute (http://ruby-doc.com/docs/ProgrammingRuby/html/tut_classes.html#UC) so that you can make this call while looking like an assignment, like `self.level = 3`.

File: lib/logdna.rb [hunk 35]
```diff
@@ -82,78 +59,68 @@ def level=(value)
       @level = value
     end
 
-    def log(msg=nil, opts={})
-      loggerExist?
-      message = msg
-      message = yield if msg.nil? && block_given?
```
Comment (dchai76) [inline lib/logdna.rb]: You got rid of the yield... why?

File: lib/logdna.rb [hunk 36]
```diff
@@ -82,78 +59,68 @@ def level=(value)
...
 
     Resources::LOG_LEVELS.each do |lvl|
       name = lvl.downcase
 
-      define_method name do |msg=nil, opts={}, &block|
-        self.log(msg, opts.merge({
-          level: lvl,
-        }), &block)
+      define_method name do |msg = nil, opts = {}, &block|
+        log(msg, opts.merge(
```
Comment (dchai76) [inline lib/logdna.rb]: I'm confused by 1/ Why it's not `self.log` anymore and 2/ (related to the above), why we're passing in a block when `log` doesn't use it.

File: lib/logdna.rb [hunk 37]
```diff
@@ -82,78 +59,68 @@ def level=(value)
...
+                   level: lvl
+                 ), &block)
       end
 
       define_method "#{name}?" do
-        return Resources::LOG_LEVELS[self.level] == lvl if self.level.is_a? Numeric
-        self.level == lvl
+        return Resources::LOG_LEVELS[level] == lvl if level.is_a? Numeric
+
+        assign_level == lvl
```
Comment (dchai76) [inline lib/logdna.rb]: This shows where your rename doesn't make sense; `assign_level` is a method, this line doesn't make sense. This should be checking the level var.

File: lib/logdna/client.rb [hunk 38]
```diff
@@ -1,164 +1,132 @@
...
-        }.to_json
+      @request.body = {
+        e: "ls",
+        ls: @buffer
+      }.to_json
 
-        @@request.body = real
-        @response = Net::HTTP.start(@uri.hostname, @uri.port, use_ssl: @uri.scheme == 'https') do |http|
-          http.request(@@request)
+      handle_excpetion = lambda do |message|
```
Comment (dchai76) [inline lib/logdna/client.rb]: Fix the spelling here and in every reference please

File: lib/logdna/client.rb [hunk 39]
```diff
@@ -1,164 +1,132 @@
...
+      ensure
+        @buffer.clear
       end
     end
 
-    def exitout()
-      check_side_buffer
-      if @messages.any?
-        flush()
+    def flush
```
Comment (dchai76) [inline lib/logdna/client.rb]: So correct me if I'm wrong but this seems weird to me - the @flush_scheduled is outside the lock, so it feels like you can get weird behavior. Isn't it only flush that needs to be locked? Can we just `synchronize` this whole section? Since we have `@flush_scheduled` if feels like that takes care of the lock waiting you're addressing. I'm also somewhat confused as to why we need locking on the side messages. Maybe explain to me.

File: lib/logdna/client.rb (hunk 31 above)
Comment (dchai76) [reply lib/logdna/client.rb]: But how do you know that that's why we didn't get a success response? What if there's some other error?

Comment (vilyapilya) [reply lib/logdna/client.rb]: I know it does not look like an ideal handling. But testing how this mechanism works in Ruby, I found this to be the best approach. Because Ruby does not really work well with general exception handling. I tested it with writing just rescue, and switching off the connection and providing a bad ingestion key.  The rescue part did not catch it. But it does catch if the specific exception names are defined.  I can probably call `handle_excpetion` lambda in the second (`!@response.is_a?(Net::HTTPSuccess)`)  condition.

File: lib/logdna/client.rb (hunk 39 above)
Comment (vilyapilya) [reply lib/logdna/client.rb]: We need to lock side_messages bc some other threads can modify it and can have weird results.

Comment (vilyapilya) [reply lib/logdna/client.rb]: Regarding other points of your comment, I think you're probably right.

Comment (vilyapilya) [reply lib/logdna/client.rb]: Actually, if I change it in this way, side-messages lock will be redundant.

Comment (vilyapilya) [reply lib/logdna/client.rb]: No, I still need the lock for side-messages which has to be short. Because I am trying to optimize here the lock usage. If the client receives the messages but some other thread occupied the buffer, we can do the following: 
1. Wait for the lock release. Which can take a long time bc it might not be release until the request is done. We're holding the lock for the whole request process
2. Write those message in to the side-messages. Which are always protected by a short lock. (@side-messaged lock should be taken only for adding/removing data from side-messages). 

So choosing the number 2, I optimize the performance. Since we have to keep @lock for a long time, let's not block other threads from giving the messages away to the lib. and the lib will take care of them later.

Comment (vilyapilya) [reply lib/logdna/client.rb]: I don't think `schedule_flush`  should be inside any locks. But I'll move the `@flush_scheduled` flag switching into locks to prevent weird behavior.

File: lib/logdna.rb [hunk 40]
```diff
@@ -82,78 +59,66 @@ def level=(value)
...
     end
 
     Resources::LOG_LEVELS.each do |lvl|
       name = lvl.downcase
 
-      define_method name do |msg=nil, opts={}, &block|
-        self.log(msg, opts.merge({
-          level: lvl,
-        }), &block)
+      define_method name do |msg = nil, opts = {}|
```
Comment (dchai76) [inline lib/logdna.rb]: So this is the part I wanted you to understand. What this is (or was) doing is defining methods for each log level `error', `warn`, `debug`, etc. And in doing so, it's mimicking the methods that the Ruby Logger has: https://ruby-doc.org/stdlib-2.4.0/libdoc/logger/rdoc/Logger.html#method-i-info. If you look at all of those methods (I linked to `info`, but it's there for all of them), you'll see that they all take a `&block` that `Evaluates to the message to log.`. You're removing the functionality by removing the block handling in `log`. If you want to do that we can, but note that it does not deviate from what Ruby's Logger looks like.

File: lib/logdna/client.rb (hunk 38 above)
Comment (vilyapilya) [reply lib/logdna/client.rb]: done

File: lib/logdna.rb (hunk 40 above)
Comment (vilyapilya) [reply lib/logdna.rb]: The original code does not replicate the Ruby Logger behavior. I'll make a change so it will d replicate it

Comment (vilyapilya) [reply lib/logdna.rb]: Actually, it makes sense why the original code does not fully replicate the Ruby Logger behavior. I have restored it.

File: lib/logdna.rb [hunk 41]
```diff
@@ -82,78 +59,69 @@ def level=(value)
...
-      message = msg
-      message = yield if msg.nil? && block_given?
-      @response = @@client.buffer(message, default_opts.merge(opts).merge({
-            timestamp: (Time.now.to_f * 1000).to_i
-        }))
-      'Saved'
+    def log(message = nil, opts = {})
+      if message.nil? && block_given?
+        message = yield
+      elsif message.nil?
```
Comment (dchai76) [inline lib/logdna.rb]: Might want to make this just `if message.nil?` to detect if the block provided is wrong.

File: lib/logdna.rb [hunk 42]
```diff
@@ -82,78 +59,69 @@ def level=(value)
...
-        }), &block)
+      define_method name do |msg = nil, opts = {}, &block|
+        self.log(msg, opts.merge(
+                        level: lvl
+                      ), &block)
       end
 
       define_method "#{name}?" do
-        return Resources::LOG_LEVELS[self.level] == lvl if self.level.is_a? Numeric
+        return Resources::LOG_LEVELS[level] == lvl if level.is_a? Numeric
```
Comment (dchai76) [inline lib/logdna.rb]: Did this work? I can't see where `level` is defined.

File: lib/logdna.rb [hunk 43]
```diff
@@ -82,78 +59,69 @@ def level=(value)
...
-      return false
+      false
     end
 
-
     def close
-      if defined? @@client and !@@client.nil?
-          @@client.exitout()
-      end
+      @client.exitout if defined? @client && !@client.nil?
```
Comment (dchai76) [inline lib/logdna.rb]: Although optional, I prefer the style of always including parens on function calls for better readability.

File: lib/logdna.rb [hunk 44]
```diff
@@ -82,78 +59,69 @@ def level=(value)
...
     end
 
-    at_exit do
-      if defined? @@client and !@@client.nil?
-          @@client.exitout()
-      end
+    def at_exit
```
Comment (dchai76) [inline lib/logdna.rb]: Wait, `at_exit` is something you should be calling - it's the block that executes on program exit - not something you should be defining. https://apidock.com/ruby/Kernel/at_exit

File: lib/logdna/client.rb (hunk 38 above)
Comment (dchai76) [reply lib/logdna/client.rb]: You said done... you didn't fix the spelling here and in every reference. Please do.

File: lib/logdna/client.rb [hunk 45]
```diff
@@ -1,164 +1,134 @@
...
+      if @lock.try_lock
+        @flush_scheduled = false
+        return if @buffer.empty? && @side_messages.empty?
```
Comment (dchai76) [inline lib/logdna/client.rb]: This seems wrong. You're returning without releasing the `@lock`. I think this could happen if 2 threads schedule a flush, one gets there first so empties the buffer/side_messages, then this thread gets here and never unlocks.

File: lib/logdna/client.rb (hunk 31 above)
Comment (dchai76) [reply lib/logdna/client.rb]: This isn't an exception though, it's a non-success response. I'd consider a retry in this case - there's no harm.

File: logdna.gemspec [hunk 46]
```diff
@@ -1,4 +1,5 @@
 # coding: utf-8
+
```
Comment (dchai76) [inline logdna.gemspec]: Prefer not having whitespace-only changes to files.

File: lib/logdna.rb (hunk 41 above)
Comment (vilyapilya) [reply lib/logdna.rb]: good point

File: lib/logdna.rb (hunk 42 above)
Comment (vilyapilya) [reply lib/logdna.rb]: it did work but I'll change to self.level. I did not notice it

File: lib/logdna/client.rb (hunk 38 above)
Comment (vilyapilya) [reply lib/logdna/client.rb]: I started to wonder whether I have dyslexia.

File: lib/logdna/client.rb (hunk 45 above)
Comment (vilyapilya) [reply lib/logdna/client.rb]: good point. Thanks

**Comment by vilyapilya** [comment]
I think "enhancement" is the wrong label here.

File: lib/logdna/client.rb [hunk 47]
```diff
@@ -1,164 +1,137 @@
...
-      return if msg.nil?
-      msg = encode_message(msg)
-
-      if @lock.locked?
-        @side_buffer.write(msg)
-        @side_messages.push(message_hash(msg, opts))
-        return
+    def send_request
+      @side_message_lock.synchronize do
+        @buffer.concat(@side_messages)
```
Comment (dchai76) [inline lib/logdna/client.rb]: I think we should leave a comment here that this function must be called with a claimed `@lock` (and/or even explicitly check for that in the code). Otherwise when I read this I'm confused why we're modifying `@buffer` without looking at `@lock`.

File: lib/logdna/client.rb [hunk 48]
```diff
@@ -1,164 +1,137 @@
...
 
-        # don't kill @task if this was executed from self.buffer
-        # don't kill @task if there are queued messages
-        unless @buffer_over_limit || @messages.any? || @task.nil?
-          @task.shutdown
-          @task.kill
+    def flush
+      if @lock.try_lock
+        @flush_scheduled = false
+        if @buffer.empty? && @side_messages.empty?
```
Comment (dchai76) [inline lib/logdna/client.rb]: Alternatively, might want to reverse the logic so:
```
if @buffer.any? || @side_messages.any?
  send_request
end
@lock.unlock
```

That way there's just a single @lock.unlock call.

File: .gitignore [hunk 49]
```diff
@@ -9,3 +9,5 @@
 /tmp/
 *.gem
 .DS_Store
+/bin/
+package-lock.json
```
Comment (dchai76) [inline .gitignore]: Ack no, this is back. Remove.
